### ISSUE TITLE
error in @wraps(f)

### ISSUE THREAD

**Issue by simplew2011** [Nuitka/Nuitka#2624]
```
1.8.4
Commercial: None
Python: 3.8.5 (default, Sep  4 2020, 07:30:14) 
Flavor: Anaconda Python
Executable: /home/wzp/anaconda3/envs/python3.8/bin/python
OS: Linux
Arch: x86_64
Distribution: Ubuntu (based on Debian) 18.04.6
Version C compiler: /usr/bin/gcc (gcc).
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
```

```
git clone https://github.com/alibaba/data-juicer.git
cd path/to/data-juicer

(add a line in [setup.py](https://github.com/alibaba/data-juicer/blob/main/setup.py#L80))
build_with_nuitka=True,

# build whl
python setup.py bdist_nuitka

# install nuitka whl
pip install dist/py_data_juicer-0.1.2-cp38-cp38-linux_x86_64.whl

# test whl

(set dataset_path in absolute path:  https://github.com/alibaba/data-juicer/blob/main/configs/demo/analyser.yaml#L5)

dj-process --config /home/wzp/code/LLMData/open_source/data-juicer/configs/demo/analyser.yaml

```

error occur：
- https://github.com/alibaba/data-juicer/blob/main/data_juicer/core/data.py#L34
```
    def wrap_nested_structure(*args, **kargs):
        wrapped_args = [nested_obj_factory(arg) for arg in args]
        wrapped_kargs = {
            k: nested_obj_factory(arg)
            for k, arg in kargs.items()
        }
        return wrapped_args, nested_obj_factory(wrapped_kargs)

    @wraps(f)
    def wrapped_f(*args, **kargs):
        args, kargs = wrap_nested_structure(*args, **kargs)
        # to ensure the args passing to the final calling of f can be nested,
        # in case of deeper-order wrapper funcs de-wrap this nesting behavior
        args = [
            wrap_func_with_nested_access(arg) if callable(arg) else arg
            for arg in args
        ]
        kargs = {
            k: (wrap_func_with_nested_access(arg) if callable(arg) else arg)
            for (k, arg) in kargs.items()
        }
        return f(*args, **kargs)

    return wrapped_f
```

```
2023-12-25 16:51:09 | WARNING  | data_juicer.utils.fingerprint_utils:98 - Parameter 'function'=<compiled_function unify_format.<locals>.non_empty_text at 0x7fe66a781440> of the                          transform datasets.arrow_dataset.Dataset._map_single couldn't be hashed properly,                          a random hash was used instead. Make sure your                          transforms and parameters are serializable with                          pickle or dill for the dataset fingerprinting and                          caching to work. If you reuse this transform, the                          caching mechanism will consider it to be different                          from the previous calls and recompute everything.                          This warning is only showed once. Subsequent hashing                          failures won't be showed.
2023-12-25 16:51:09 | ERROR    | __main__:8 - An error has been caught in function '<module>', process 'MainProcess' (24219), thread 'MainThread' (140630909462336):
Traceback (most recent call last):

> File "/home/wzp/anaconda3/envs/3.8/bin/dj-process", line 8, in <module>
    sys.exit(main())
    │   │    └ <function main at 0x7fe72da74310>
    │   └ <built-in function exit>
    └ <module 'sys' (built-in)>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/data_juicer/tools/process_data.py", line 15, in main
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/data_juicer/core/executor.py", line 102, in run
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/data_juicer/format/mixture_formatter.py", line 134, in load_dataset
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/data_juicer/format/formatter.py", line 77, in load_dataset
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/data_juicer/format/formatter.py", line 211, in unify_format
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/data_juicer/core/data.py", line 228, in filter
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/datasets/arrow_dataset.py", line 528, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
         │                                 │    │      │       └ {'num_proc': 4, 'fn_kwargs': {'target_keys': ['text']}, 'new_fingerprint': '17216009ac692663'}
         │                                 │    │      └ (<compiled_function unify_format.<locals>.non_empty_text at 0x7fe66a781440>,)
         │                                 │    └ Dataset({
         │                                 │          features: ['text', 'meta'],
         │                                 │          num_rows: 6
         │                                 │      })
         │                                 └ <function Dataset.filter at 0x7fe71270bc10>
         └ typing.Union
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/datasets/fingerprint.py", line 511, in wrapper
    out = func(dataset, *args, **kwargs)
          │    │         │       └ {'num_proc': 4, 'fn_kwargs': {'target_keys': ['text']}, 'new_fingerprint': '17216009ac692663'}
          │    │         └ (<compiled_function unify_format.<locals>.non_empty_text at 0x7fe66a781440>,)
          │    └ Dataset({
          │          features: ['text', 'meta'],
          │          num_rows: 6
          │      })
          └ <function Dataset.filter at 0x7fe71270bb80>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/datasets/arrow_dataset.py", line 3531, in filter
    indices = self.map(
              │    └ <compiled_function NestedDataset.map at 0x7fe670a5f740>
              └ Dataset({
                    features: ['text', 'meta'],
                    num_rows: 6
                })
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/data_juicer/core/data.py", line 180, in map
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/datasets/arrow_dataset.py", line 563, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
         │                                 │    │      │       └ {'function': <compiled_function wrap_func_with_nested_access.<locals>.wrapped_f at 0x7fe66a781540>, 'with_indices': True, 'fe...
         │                                 │    │      └ ()
         │                                 │    └ Dataset({
         │                                 │          features: ['text', 'meta'],
         │                                 │          num_rows: 6
         │                                 │      })
         │                                 └ <function Dataset.map at 0x7fe71270b940>
         └ typing.Union
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/datasets/arrow_dataset.py", line 528, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
         │                                 │    │      │       └ {'function': <compiled_function wrap_func_with_nested_access.<locals>.wrapped_f at 0x7fe66a781540>, 'with_indices': True, 'fe...
         │                                 │    │      └ ()
         │                                 │    └ Dataset({
         │                                 │          features: ['text', 'meta'],
         │                                 │          num_rows: 6
         │                                 │      })
         │                                 └ <function Dataset.map at 0x7fe71270b8b0>
         └ typing.Union
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/datasets/arrow_dataset.py", line 3097, in map
    for rank, done, content in iflatmap_unordered(
        │                      └ <function iflatmap_unordered at 0x7fe712cd35e0>
        └ 3
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/datasets/utils/py_utils.py", line 1377, in iflatmap_unordered
    [async_result.get() for async_result in async_results]
                                            └ [<multiprocess.pool.ApplyResult object at 0x7fe66a6f9a60>, <multiprocess.pool.ApplyResult object at 0x7fe66a6def40>, <multipr...
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/datasets/utils/py_utils.py", line 1377, in <listcomp>
    [async_result.get() for async_result in async_results]
     │            │         └ <multiprocess.pool.ApplyResult object at 0x7fe66a6f9a60>
     │            └ <function ApplyResult.get at 0x7fe712cd09d0>
     └ <multiprocess.pool.ApplyResult object at 0x7fe66a6f9a60>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/multiprocess/pool.py", line 771, in get
    raise self._value
          │    └ PicklingError("Can't pickle <compiled_function wrap_func_with_nested_access.<locals>.wrapped_f at 0x7fe66a781540>: it's not f...
          └ <multiprocess.pool.ApplyResult object at 0x7fe66a6f9a60>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/multiprocess/pool.py", line 537, in _handle_tasks
    put(task)
    │   └ None
    └ <bound method _ConnectionBase.send of <multiprocess.connection.Connection object at 0x7fe66a74a430>>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/multiprocess/connection.py", line 209, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
    │    │           │               │     └ (0, 0, <function _write_generator_to_queue at 0x7fe712cd3550>, (<AutoProxy[Queue] object, typeid 'Queue' at 0x7fe66a735220>, ...
    │    │           │               └ <classmethod object at 0x7fe714f8c160>
    │    │           └ <class 'multiprocess.reduction.ForkingPickler'>
    │    └ <function Connection._send_bytes at 0x7fe712ccdd30>
    └ <multiprocess.connection.Connection object at 0x7fe66a74a430>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/multiprocess/reduction.py", line 54, in dumps
    cls(buf, protocol, *args, **kwds).dump(obj)
    │   │    │          │       │          └ (0, 0, <function _write_generator_to_queue at 0x7fe712cd3550>, (<AutoProxy[Queue] object, typeid 'Queue' at 0x7fe66a735220>, ...
    │   │    │          │       └ {}
    │   │    │          └ ()
    │   │    └ None
    │   └ <_io.BytesIO object at 0x7fe6696983b0>
    └ <class 'multiprocess.reduction.ForkingPickler'>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/dill/_dill.py", line 498, in dump
    StockPickler.dump(self, obj)
    │            │    │     └ (0, 0, <function _write_generator_to_queue at 0x7fe712cd3550>, (<AutoProxy[Queue] object, typeid 'Queue' at 0x7fe66a735220>, ...
    │            │    └ <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>
    │            └ <function _Pickler.dump at 0x7fe72d97a430>
    └ <class 'pickle._Pickler'>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 485, in dump
    self.save(obj)
    │    │    └ (0, 0, <function _write_generator_to_queue at 0x7fe712cd3550>, (<AutoProxy[Queue] object, typeid 'Queue' at 0x7fe66a735220>, ...
    │    └ <function _Pickler.save at 0x7fe72d97a670>
    └ <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 558, in save
    f(self, obj)  # Call unbound method with explicit self
    │ │     └ (0, 0, <function _write_generator_to_queue at 0x7fe712cd3550>, (<AutoProxy[Queue] object, typeid 'Queue' at 0x7fe66a735220>, ...
    │ └ <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>
    └ <function _Pickler.save_tuple at 0x7fe72d97ad30>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 899, in save_tuple
    save(element)
    │    └ (<AutoProxy[Queue] object, typeid 'Queue' at 0x7fe66a735220>, <function Dataset._map_single at 0x7fe71270ba60>, {'shard': Dat...
    └ <bound method _Pickler.save of <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 558, in save
    f(self, obj)  # Call unbound method with explicit self
    │ │     └ (<AutoProxy[Queue] object, typeid 'Queue' at 0x7fe66a735220>, <function Dataset._map_single at 0x7fe71270ba60>, {'shard': Dat...
    │ └ <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>
    └ <function _Pickler.save_tuple at 0x7fe72d97ad30>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 884, in save_tuple
    save(element)
    │    └ {'shard': Dataset({
    │          features: ['text', 'meta'],
    │          num_rows: 2
    │      }), 'function': <compiled_function wrap_func_with_nested_...
    └ <bound method _Pickler.save of <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 558, in save
    f(self, obj)  # Call unbound method with explicit self
    │ │     └ {'shard': Dataset({
    │ │           features: ['text', 'meta'],
    │ │           num_rows: 2
    │ │       }), 'function': <compiled_function wrap_func_with_nested_...
    │ └ <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>
    └ <function save_module_dict at 0x7fe714f61040>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/site-packages/dill/_dill.py", line 990, in save_module_dict
    StockPickler.save_dict(pickler, obj)
    │            │         │        └ {'shard': Dataset({
    │            │         │              features: ['text', 'meta'],
    │            │         │              num_rows: 2
    │            │         │          }), 'function': <compiled_function wrap_func_with_nested_...
    │            │         └ <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>
    │            └ <function _Pickler.save_dict at 0x7fe72d97aee0>
    └ <class 'pickle._Pickler'>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 969, in save_dict
    self._batch_setitems(obj.items())
    │    │               │   └ <method 'items' of 'dict' objects>
    │    │               └ {'shard': Dataset({
    │    │                     features: ['text', 'meta'],
    │    │                     num_rows: 2
    │    │                 }), 'function': <compiled_function wrap_func_with_nested_...
    │    └ <function _Pickler._batch_setitems at 0x7fe72d97af70>
    └ <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 995, in _batch_setitems
    save(v)
    │    └ <compiled_function wrap_func_with_nested_access.<locals>.wrapped_f at 0x7fe66a781540>
    └ <bound method _Pickler.save of <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 587, in save
    self.save_global(obj, rv)
    │    │           │    └ 'wrap_func_with_nested_access.<locals>.wrapped_f'
    │    │           └ <compiled_function wrap_func_with_nested_access.<locals>.wrapped_f at 0x7fe66a781540>
    │    └ <function _Pickler.save_global at 0x7fe72d97c160>
    └ <multiprocess.reduction.ForkingPickler object at 0x7fe66a714c10>
  File "/home/wzp/anaconda3/envs/3.8/lib/python3.8/pickle.py", line 1068, in save_global
    raise PicklingError(
          └ <class '_pickle.PicklingError'>

_pickle.PicklingError: Can't pickle <compiled_function wrap_func_with_nested_access.<locals>.wrapped_f at 0x7fe66a781540>: it's not found as data_juicer.core.data.wrap_func_with_nested_access.<locals>.wrapped_f
```

**Comment by simplew2011** [comment]
`python setup.py bdist_wheel` is ok

**Comment by kayhayen** [comment]
Any reason why you are not using the 1.9 release for it? The dill plugin probably would help, but it doesn't work in module mode yet, however I have this working on factory branch. Since this is pure `pickle` for multiprocess, it probably can have a similar solution, with registering the compiled types there too. Functions that are local to other functions are harder to restore, and I am now not sure anymore, if closure works, but I think it does for dill, but this clearly needs a more minimal reproducer.

**Comment by simplew2011** [comment]
- update to nuitka-1.9

- and add `dill-compat` in setup.py

```
setup(
    ...
    command_options={
        'nuitka': {
            '--enable-plugin': "dill-compat",
        },
    },
)
```

- build whl：python setup.py bdist_nuitka
- install whl ok：pip install dist/xx.whl
- use library occur error: Segmentation fault (core dumped)

**Comment by simplew2011** [comment]
- use factory branch
- git clone https://github.com/Nuitka/Nuitka.git
- git checkout -b factory
- python setup.py install
- python setup.py bdist_nuitka
- error occur 

```
2.0rc6
Commercial: None
Python: 3.8.5 (default, Sep  4 2020, 07:30:14) 
Flavor: Anaconda Python
Executable: /home/wzp/anaconda3/envs/python3.8/bin/python
OS: Linux
Arch: x86_64
Distribution: Ubuntu (based on Debian) 18.04.6
Version C compiler: /usr/bin/gcc (gcc 7).
```

```
subprocess.CalledProcessError: Command '['/home/wzp/anaconda3/envs/python3.8/bin/python', '-m', 'nuitka', '--module', '--enable-plugin=pylint-warnings', '--output-dir=/home/wzp/code/LLMData/VastDataSense/build/lib', '--nofollow-import-to=*.tests', '--remove-output', '--include-package=datasense', '--enable-plugin=dill-compat', '/home/wzp/code/LLMData/VastDataSense/build/lib/datasense']' returned non-zero exit status 1.

```

```
<?xml version='1.0' encoding='utf8'?>
<nuitka-compilation-report nuitka_version="2.0rc6" nuitka_commercial_version="not installed" completion="exception">
  <exception exception_type="FileNotFoundError" exception_value="[Errno 2] No such file or directory: '/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/plugins/standard/DillPlugin/DillPlugin.c'">
Traceback (most recent call last):
  File "/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/MainControl.py", line 1116, in main
    _main()
  File "/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/MainControl.py", line 974, in _main
    result, scons_options = compileTree()
  File "/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/MainControl.py", line 886, in compileTree
    Plugins.writeExtraCodeFiles(onefile=False)
  File "/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/plugins/Plugins.py", line 1279, in writeExtraCodeFiles
    for filename, source_code in Plugins._getExtraCodeFiles(onefile).items():
  File "/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/plugins/Plugins.py", line 1251, in _getExtraCodeFiles
    value = plugin.getExtraCodeFiles()
  File "/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/plugins/standard/DillPlugin.py", line 51, in getExtraCodeFiles
    return {"DillPlugin.c": self.getPluginDataFileContents("DillPlugin.c")}
  File "/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/plugins/PluginBase.py", line 957, in getPluginDataFileContents
    return getFileContents(
  File "/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/utils/FileOperations.py", line 760, in getFileContents
    with openTextFile(filename, mode, encoding=encoding) as f:
  File "/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/utils/FileOperations.py", line 793, in openTextFile
    return open(filename, mode)  # pylint: disable=unspecified-encoding
FileNotFoundError: [Errno 2] No such file or directory: '/home/wzp/anaconda3/envs/python3.8/lib/python3.8/site-packages/Nuitka-2.0rc6-py3.8-linux-x86_64.egg/nuitka/plugins/standard/DillPlugin/DillPlugin.c'
```

**Comment by kayhayen** [comment]
That shouldn't happen on factory anymore, and 1.9.6 is also fixed, but I didn't say that the dill plugin alone is what you need, but also some extra work, that is not there yet.

It seems a minimal example is not forthcoming, or is it?

**Comment by simplew2011** [comment]
The following minimum example can be reproduced：

```
# step 1
git clone https://github.com/alibaba/data-juicer.git
cd path/to/data-juicer

# step 2
(add a line in [setup.py](https://github.com/alibaba/data-juicer/blob/main/setup.py#L80))
build_with_nuitka=True,

# step 3
# build whl
python setup.py bdist_nuitka

# step 4
# install nuitka whl
pip install dist/py_data_juicer-0.1.2-cp38-cp38-linux_x86_64.whl

# step 5
# test whl
(set dataset_path in absolute path:  https://github.com/alibaba/data-juicer/blob/main/configs/demo/analyser.yaml#L5)

dj-process --config data-juicer/configs/demo/analyser.yaml
```

**Comment by kayhayen** [comment]
So, wait, this is actually `dill` and module mode, that got fixed with 1.9.6 hotfix. I was thinking it's something different. Just retry with latest stable release then.

**Comment by kayhayen** [comment]
Module mode wasn't supported by the dill plugin previously, and I misread the callstack to be something different than dill, but when I now was asking for somebody to maybe extract a minimal example, it seems that was dill all along.

On a side note, a commercial user effectively paid for this by becoming priority subscriber. Let me know how well it works for you, since if there is missing things for dill, that would be great to know.
